### PR TITLE
[WIP/POC]: Base module approach

### PR DESCRIPTION
**Pull request by Milerius**
## Description

WIP/POC  approach for internal modularity for #2483 

## How to test

Run c++ unit tests

## Types of changes

Improvements

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -42,6 +42,8 @@ include(ExternalProject)
 # Dependencies
 include(cmake/Protobuf.cmake)
 
+add_subdirectory(modules)
```
[the remark this answers is not in the document]
Comment (Milerius) [reply CMakeLists.txt:45]: A few things, the sources folder uses globs which makes handling the folders particularly complicated with CMake, once all the modules have been created, we can rename the folder to `src` personally I have a preference for modules (C++ modules) / interfaces (c API)

File: modules/base/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+add_library(tw_base OBJECT)
+target_sources(tw_base PRIVATE Data.cpp)
```
[the remark this answers is not in the document]
Comment (Milerius) [reply modules/base/CMakeLists.txt:2]: Glob is historically known as harmful on CMake: https://discourse.cmake.org/t/is-glob-still-considered-harmful-with-configure-depends/808 - I suggest the new approach to note using it. It's also the reason it's hard to have the new modules under the `src` folder

File: swift/project.yml [hunk 3]
```diff
@@ -4,7 +4,7 @@ options:
   createIntermediateGroups: true
 settings:
   base:
-    HEADER_SEARCH_PATHS: $(SRCROOT)/wallet-core ${SRCROOT}/trezor-crypto/crypto
+    HEADER_SEARCH_PATHS: $(SRCROOT)/wallet-core ${SRCROOT}/trezor-crypto/crypto  $(SRCROOT)/wallet-core-modules/base
```
[the remark this answers is not in the document]
Comment (Milerius) [reply swift/project.yml:7]: Unfortunately, I didn't find an alternative way on project.yml - maybe I can automate that adjustment with a script?

File: CMakeLists.txt [hunk 4]
```diff
@@ -57,6 +59,8 @@ else ()
     target_link_libraries(TrustWalletCore PUBLIC ${PROJECT_NAME}_INTERFACE PRIVATE TrezorCrypto protobuf Boost::boost)
 endif ()
 
+target_link_libraries(TrustWalletCore PUBLIC tw::base)
```
[the remark this answers is not in the document]
Comment (Milerius) [reply CMakeLists.txt]: Yes, I agree. We can rename it later!

**Comment by hewigovens** [comment]
Let's merge header path changes for now? reopen this pr when we start to do module thingy

**Pull request closed without merging**